src/ripple/app/misc/SHAMapStoreImp.cpp

```diff
@@ -567,24 +500,64 @@ SHAMapStoreImp::makeBackendRotating(std::string path)
 }
 
 void
-SHAMapStoreImp::clearSql(
-    DatabaseCon& database,
-    LedgerIndex lastRotated,
-    std::string const& minQuery,
-    std::string const& deleteQuery)
+SHAMapStoreImp::clearLedgerSql(LedgerIndex lastRotated)
 {
     assert(deleteInterval_);
     LedgerIndex min = std::numeric_limits<LedgerIndex>::max();
+    const std::string TableName = "Ledgers";
 
     {
-        boost::optional<std::uint64_t> m;
+        JLOG(journal_.trace()) << "Begin: Look up lowest value of: " << TableName;
+        auto m = app_.getSQLInterface().getMinLedgerSeq();
+        JLOG(journal_.trace()) << "End: Look up lowest value of: " << TableName;
+        if (!m)
+            return;
+        min = *m;
+    }
+
+    if (min > lastRotated || health() != Health::ok)
+        return;
+    if (min == lastRotated)
+    {
+        // Micro-optimization mainly to clarify logs
+        JLOG(journal_.trace()) << "Nothing to delete from " << TableName;
+        return;
+    }
+
+    JLOG(journal_.debug()) << "start deleting in: " << TableName
+                           << " from " << min << " to " << lastRotated;
+    while (min < lastRotated)
+    {
+        min = std::min(lastRotated, min + deleteBatch_);
         JLOG(journal_.trace())
-            << "Begin: Look up lowest value of: " << minQuery;
-        {
-            auto db = database.checkoutDb();
-            *db << minQuery, soci::into(m);
-        }
-        JLOG(journal_.trace()) << "End: Look up lowest value of: " << minQuery;
+            << "Begin: Delete up to " << deleteBatch_
+            << " rows with LedgerSeq < " << min
+            << " using deleting in: " << TableName;
+        app_.getSQLInterface().deleteBeforeLedgerSeq(min);
+        JLOG(journal_.trace())
+            << "End: Delete up to " << deleteBatch_ << " rows with LedgerSeq < "
+            << min << " using deleting in: " << TableName;
+        if (health())
+            return;
+        if (min < lastRotated)
+            std::this_thread::sleep_for(backOff_);
+        if (health())
+            return;
+    }
+    JLOG(journal_.debug()) << "finished deleting in: " << TableName;
+}
+
+void
+SHAMapStoreImp::clearTransactionSql(LedgerIndex lastRotated)
+{
```

Comment:
The functions `clearLedgerSql`, `clearTransactionSql`, and `clearAccountTransactionSql` are all almost identical. I'm fine with the interface as-is, but their implementation should take advantage of this commonality. Can we add a helper function that all three call?

Reply:
Done.